datadog_checks_base/datadog_checks/base/utils/db/statement_metrics.py

```
@@ -0,0 +1,105 @@
+# (C) Datadog, Inc. 2020-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+from datadog_checks.base import is_affirmative
+
+try:
+    import datadog_agent
+except ImportError:
+    from ..stubs import datadog_agent
+
+
+class StatementMetrics:
+    """
+    This class supports normalized statement-level metrics, which are collected from the database's
+    statistics tables, ex:
+
+        - Postgres: pg_stat_statements
+        - MySQL: performance_schema.events_statements_summary_by_digest
+        - Oracle: V$SQLAREA
+        - SQL Server: sys.dm_exec_query_stats
+        - DB2: mon_db_summary
+
+    These tables are monotonically increasing.
+    """
+    def __init__(self):
+        self.previous_statements = dict()
+
+    @staticmethod
+    def is_enabled():
+        return is_affirmative(datadog_agent.get_config('deep_database_monitoring'))
+
+    def _compute_derivative_rows(self, rows, metrics, key):
+        """
+        Compute the first derivative of column-based metrics for a given set of rows. This also resets the
+        statement cache so should only be called once per check run.
+
+        - **rows** (_List[dict]_) - rows from current check run
+        - **previous** (_List[dict]_) - rows from the previous check run
+        - **metrics** (_List[str]_) - the metrics to compute for each row
+        - **key** (_callable_) - function for an ID which uniquely identifies a row across runs
+        """
+        result = []
+        new_cache = {}
+        for row in rows:
+            row_key = key(row)
+            new_cache[row_key] = row
+            prev = self.previous_statements.get(row_key)
+            if prev is None:
+                continue
+            if any([row[k] - prev[k] < 0 for k in metrics]):
```

Since this is meant to be reused, maybe better be safe and guard against missing key errors blowing up here. Better miss some metrics than have the whole check crash.

---

In general I would prefer a loud error over silently dropping metrics. Is there a way in which this could fail due to some reason other than programmer error? If the column does not exist, then the query should fail upstream.

---

Yeah it could fail if:
* there's a typo in a single metric in the agent config, so we're not using the defaults. In that case as a user I'd rather the whole thing not crash just because there's a typo in one of the metric keys in `query_metric_limits` 
* in postgres it's the same, there's differences in columns between versions, and if you take one `query_metric_limits` config and point it at an older version, I think it's better that the new metrics are just ignored than to have the whole check crash? 

We could log a warning?

---

Fair points, will update with a warning.